### PR TITLE
Generate purchase token and append to all url's leading to checkout

### PR DESCRIPTION
**Pull request by elliottprogrammer**
This PR generates and adds a purchase token to outgoing urls leading to checkout. ie- "Upgrade" button CTA's and the "Plans" link. For use in the [Logged Out Visitor Checkout](pbtFFM-OD-p2) project.

**The idea and purpose behind this PR:**
When a Jetpack connected owner is logged into their wp-admin dashboard, a unique token is generated and saved to the wp_options table.  When a user (who is currently not logged in to WordPress.com) goes to upgrade/purchase a plan or product, this `purchasetoken` will be sent along to the Wordpress.com userless checkout page, where upon purchase, WordPress.com will request the site to validate the purchase token before completing the purchase.  After this validation step, the `purchasetoken` will be deleted from the options table, making it so a `purchasetoken` is only validated once before a new `purchacetoken` in generated again.  The purpose of this `purchasetoken` is so that the WordPress.com userless checkout(not logged into WordPress.com) can verify that the user originated from their Jetpack enabled site and they are the connection owner and the site is a valid jetpack connected site.

#### Changes proposed in this Pull Request:

* If `purchaseToken` doesn't exist in options table, one is created and then added to Redux `initialState` property.
* A `getPurchaseToken( state)` selector is added.
* A `purchasetoken` query param is added to all "Upgrade" button urls and to the "Plans" page url.
* ~~Added xmlrpc methods to retrieve & validate the client purchase token from Jetpack Server.~~ The endpoints were added in #19924.

#### Jetpack product discussion

* [Project Thread: Logged Out Visitor Checkout](pbtFFM-OD-p2).
* Asana Task: 1200152453830945-as-1200222681647748

#### Does this pull request change what data or activity we track or use?
No, I don't think so.

#### Testing instructions:

1. Checkout this PR and run locally or run this branch using the Jetpack Beta plugin.
2. Visit the Jetpack Dashboard page and verify that the "Plans" url now contains a `purchasetoken` param & value.
3. Scroll down further and verify that all the "Upgrade" button url's contain the `purchasetoken` param & value.
.

**Screenshot:** "Plans" url should now contain a `query` param with the `purchasetoken` key/val pair.
Markup 2021-04-22 at 17 39 42
.

**Screenshot:** "Upgrade" CTA's should now contain `purchasetoken` param.
Markup 2021-04-22 at 17 34 47

### CODE REVIEW

File: projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php
Comment (rcoll) [inline projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php:198]: Should be `Jetpack_Options::update_option()` instead of just `update_option()`.

File: projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php
Comment (rcoll) [inline projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php:192]: We need to bail early in this method when the current user is not the main/master/connecting user.

File: projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js
Comment (jeherve) [inline projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js:33]: Maybe we could without lodash here, and instead rely on optional chaining?

File: projects/plugins/jetpack/class-jetpack-xmlrpc-methods.php
Comment (jeherve) [inline projects/plugins/jetpack/class-jetpack-xmlrpc-methods.php:36]: I believe we're trying to move away from the "userless" naming now. See #19734. Maybe we should avoid introducing such changes here now? cc @leogermani

File: projects/plugins/jetpack/class.jetpack.php
Comment (jeherve) [inline projects/plugins/jetpack/class.jetpack.php:7673]: @leogermani could speak to this some more, but would it make sense to use our helper methods in situations like this?
https://github.com/Automattic/jetpack/blob/51edd8fd3b366e620c82624792fa52922513b23f/projects/packages/connection/src/class-manager.php#L533-L541
https://github.com/Automattic/jetpack/blob/51edd8fd3b366e620c82624792fa52922513b23f/projects/packages/connection/src/class-manager.php#L582-L592

File: projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php
Comment (jeherve) [inline projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php:213]: Do we need the `function_exists` check here? It seems `wp_generate_password` was added to WP quite a while ago now.

File: composer.json
Comment (jeherve) [inline composer.json:11]: I don't know if this addition is necessary, since we do not use the changelogger in the root of the monorepo.

**Review by jeherve — COMMENTED** [review]
I gave it a quick first look. Hopefully it helps a bit! It looks like you'll need to run `tools/project-version.sh -u 1.13.0-alpha packages/options` to fix the failing checks.

File: projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js
Comment (leogermani) [reply projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js:33]: Yes! Please refer to p1HpG7-bRQ-p2

File: projects/plugins/jetpack/class-jetpack-xmlrpc-methods.php
Comment (leogermani) [reply projects/plugins/jetpack/class-jetpack-xmlrpc-methods.php:36]: Yes! Please refer to p1HpG7-bRQ-p2

File: projects/plugins/jetpack/class.jetpack.php
Comment (leogermani) [reply projects/plugins/jetpack/class.jetpack.php:7673]: Again, yes! `Connection\Manager` has all the methods you need to check the connection status.

File: composer.json
Comment (elliottprogrammer) [reply composer.json]: @jeherve , @leogermani , Thanks very much for reviewing and for your feedback!  Very helpful. I've made feedback corrections. 🙂

> It looks like you'll need to run `tools/project-version.sh -u 1.13.0-alpha packages/options` to fix the failing checks.

Hmm, I ran the above command but I'm still having issues with the failing checks. 🤦‍♂️ Something to do with conflicting composer.json files?  Apologies, this is my first PR to the Jetpack plugin and I'm not understanding how these checks work or how to fix them. Any help or guidance you could provide would be hugely appreciated 😊

**Comment by elliottprogrammer** [comment]
@jeherve , @leogermani , Thanks very much for reviewing and for your feedback!  Very helpful. I've made feedback corrections. 🙂

> It looks like you'll need to run `tools/project-version.sh -u 1.13.0-alpha packages/options` to fix the failing checks.

Hmm, I ran the above command but I'm still having issues with the failing checks. 🤦‍♂️ Something to do with conflicting composer.json files?  Apologies, this is my first PR to the Jetpack plugin and I'm not understanding how these checks work or how to fix them. Any help or guidance you could provide would be hugely appreciated 😊

**Comment by jeherve** [comment]
> Something to do with conflicting composer.json files? 

It seems that your branch is a bit out of date. I would recommend that you `git merge origin/master`, and then run the update command again to update the version of that package everywhere.

Let me know how that goes!

**Review by jeherve — COMMENTED** [review]
I updated your PR to solve those annoying dependency update issues. Things should be better now.

I've also added a few minor / nitpick comments below.

Beyond that, I've started a more general discussion in p1620839073091700-slack-CBG1CP4EN about the implementation of new XML-RPC methods. I wonder if we shouldn't aim to use V2 endpoints if possible. I'll let others weigh in on that :)

Edit: after discussion there, let's try to use a v2 endpoint here instead of new XML-RPC methods. Those live here:
https://github.com/Automattic/jetpack/tree/try/add-jetpack-purchase-token/projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints

Handling authentication in those endpoints for site-connected sites without a user connection is something new though. You can find out more about it here:
- p9dueE-2Kr-p2#comment-5006
- https://github.com/Automattic/jetpack/blob/0ac612fd26e7e04d6558d37fea80af15210b3fc8/projects/packages/connection/src/class-rest-authentication.php#L70-L78

We'll be happy to help with that! Don't hesitate to drop in #jetpack-crew if you have questions.

**Comment by rcanepa** [comment]
> Caution: This PR has changes that must be merged to WordPress.com
Hello elliottprogrammer! These changes need to be synced to WordPress.com - If you 're an a11n, please commandeer and confirm D61601-code works as expected before merging this PR. Once this PR is merged, please commit the changes to WP.com. Thank you!
This revision will be updated with each commit to this PR

I'm already taking care of this. WordPress.com revision will require more work since now it will use v2 endpoints instead of XML-RPC methods.

**Comment by rcanepa** [comment]
Hey @jeherve, I managed to get most of the work about replacing the XML-RPC methods done. However, requests fail with a 401. I suspect that I'm missing something auth-related (probably I'm not passing the blog token as I should). I'm also unsure about the way I'm making the request. Is using `Client::wpcom_json_api_request_as_blog` the standard way of doing that?

Here is D61601-code that complements the work being done on this PR. You can see that we call these new endpoints from `class.jetpack-purchase-token.php`.

This is the first time I do this, so bear with me :).

Thank you!

**Comment by rcanepa** [comment]
I forgot to mention that in order to test this, one can use the following URL `https://wordpress.com/checkout/jetpack/[site]/jetpack_backup_daily?site=[site]&unlinked=1&purchasetoken=3274SADasdSADASDa2` making sure that you're sandboxing `public-api.wordpress.com` and that the WP.com revision is applied to your sandbox as well. The purchase token is a fake one, because it doesn't matter at this point since we are only trying to solve the communication/auth issue.

File: projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-purchase-token.php
Comment (jeherve) [inline projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-purchase-token.php:77]: If there can only be one token, do we need to return the result as part of an array?

File: projects/plugins/jetpack/class.jetpack.php
Comment (jeherve) [inline projects/plugins/jetpack/class.jetpack.php:7652]: You'll need to update this method since we've deprecated it:
https://github.com/Automattic/jetpack/blob/master/projects/packages/connection/src/class-manager.php#L606

```suggestion
		if ( ! self::connection()->is_site_connection() ) {
```

**Review by jeherve — COMMENTED** [review]
It would be nice to have tests to provide coverage for all those new methods. You have some examples of tests for the endpoints here:
https://github.com/Automattic/jetpack/tree/try/add-jetpack-purchase-token/projects/plugins/jetpack/tests/php/core-api/wpcom-endpoints

> However, requests fail with a 401. I suspect that I'm missing something auth-related (probably I'm not passing the blog token as I should). I'm also unsure about the way I'm making the request. Is using Client::wpcom_json_api_request_as_blog the standard way of doing that?

cc @fgiannar, who is more familiar with this than I am.

File: projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-purchase-token.php
Comment (rcanepa) [reply projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-purchase-token.php:77]: I don't think there is a reason. I'll remove the array.

File: projects/plugins/jetpack/class.jetpack.php
Comment (rcanepa) [reply projects/plugins/jetpack/class.jetpack.php:7652]: Thanks. Updating this too.

**Comment by fgiannar** [comment]
Hi @rcanepa  !

I'd like to discuss a bit endpoints and authorization:

I see you chose to implement the new endpoints as `wpcom/v2` bridged endpoints.
This means that those endpoints will be available not only to Jetpack & Atomic sites, but Simple ones as well.

I don't see a reason to make the purchase token endpoints available to Simple sites, given they are targeted for remote sites only and even more the ones with site-only connections!

Instead, I'd suggest using core REST API endpoints but under the `jetpack/v4` namespace.

Regarding `authentication` as @jeherve already pointed out and you implemented, we need to make those endpoints available with site-level authentication (blog token).

I'd suggest moving this functionality to it's own PR if that's possible so it can be better tested and reviewed. If that's not possible, let's aim to provide unit tests explicitly for this part as it's very important from a security perspective.

Finally, regarding making requests from WPCOM to the corresponding endpoints, here's a one-time script I wrote for making requests from WPCOM to remote Jetpack sites using either a blog or user token: 2920a-pb/#php

Hope you find this useful!

**Comment by rcanepa** [comment]
> Hi @rcanepa !
> 
> I'd like to discuss a bit endpoints and authorization:
> 
> I see you chose to implement the new endpoints as `wpcom/v2` bridged endpoints.
> This means that those endpoints will be available not only to Jetpack & Atomic sites, but Simple ones as well.
> 
> I don't see a reason to make the purchase token endpoints available to Simple sites, given they are targeted for remote sites only and even more the ones with site-only connections!
> 
> Instead, I'd suggest using core REST API endpoints but under the `jetpack/v4` namespace.

TIL. Thank you for explaining the difference. 

I put the endpoints under the `jetpack/v4` namespace per your suggestion (I completely agree with it).

> Regarding `authentication` as @jeherve already pointed out and you implemented, we need to make those endpoints available with site-level authentication (blog token).
> 
> I'd suggest moving this functionality to it's own PR if that's possible so it can be better tested and reviewed. If that's not possible, let's aim to provide unit tests explicitly for this part as it's very important from a security perspective.

I'll be AFK tomorrow (Friday) so I'll move the endpoints to a different PR next Monday, but I had time to add some basic tests to cover them. They do not test the security aspect, though. I'll need more guidance to accomplish that – if you could point me to an example would be great.

> Finally, regarding making requests from WPCOM to the corresponding endpoints, here's a one-time script I wrote for making requests from WPCOM to remote Jetpack sites using either a blog or user token: 2920a-pb/#php

Awesome. I'll definitely that a look at that one-time script next week.

> Hope you find this useful!

You can't imagine how useful it was. Thanks a ton 🙇 !

**Comment by rcanepa** [comment]
Hey @fgiannar. I just opened a new PR (https://github.com/Automattic/jetpack/pull/19924) that adds the two `jetpack/v4` endpoints. I added some tests and I used your one-time script to test it (it worked with a site-level connection). I would appreciate it very much your review.

cc @jeherve

**Comment by rcanepa** [comment]
I removed the changes that added the two new endpoints since they are being added in #19924.

cc @jeherve

**Comment by rcanepa** [comment]
Some e2e tests are failing and throwing the following error: 

````
 ReferenceError: window is not defined

      26 | export default function getRedirectUrl( source, args = {} ) {
      27 | 	const queryVars = {};
    > 28 | 	const calypsoEnv = window.Initial_State?.calypsoEnv;
         | 	                ^
      29 |
      30 | 	if ( source.search( 'https://' ) === 0 ) {
      31 | 		const parsedUrl = new URL( source );
````

I tried to solve these errors by adding `--env=jsdom` at the end of the `"test-e2e"` script (`package.json`) but it didn't work. The issue was replaced by the following:

````
ReferenceError: page is not defined

      134 |
      135 | export async function loginToWpComIfNeeded( wpComUser, mockPlanData ) {
    > 136 | 	const login = await LoginPage.visit( page );
          | 	                                     ^
      137 | 	if ( ! mockPlanData ) {
      138 | 		await login.setSandboxModeForPayments( cookie );
      139 | 	}

      at loginToWpComIfNeeded (lib/flows/jetpack-connect.js:136:39)
      at maybePreConnect (lib/env/test-setup.js:16:8)
      at lib/env/test-setup.js:39:8
````

Is this something that can be fixed by setting `jsdom` as the testing environment?

**Comment by rcanepa** [comment]
@jeherve, is it possible to get this work merged before the code freeze? We would like to enable the new Logged Out Visitor Checkout flow as part of the new Jetpack release.

**Comment by jeherve** [comment]
> @jeherve, is it possible to get this work merged before the code freeze? We would like to enable the new Logged Out Visitor Checkout flow as part of the new Jetpack release.

I'm afraid that's not going to be possible, since we're past the code freeze at this point. Those are important changes to our flows, so I think it would be best if we could merge them to `master` early in the 9.9 release cycle and thus have a few weeks to test them well together.

**Comment by jeherve** [comment]
Adding the "Hold" label since we'll need to merge #19924 first.

**Comment by rcanepa** [comment]
Update the PR and fixed the issue that was breaking the e2e tests.

**Comment by rcanepa** [comment]
Updated the PR again since #19924 was merged.

I also made the `purchasetoken` query parameter appear only if there is a purchase token in the app state (we don't want to have something like `&purchasetoken=false` in the URL).

I had to change the HTTP method associated with the endpoint that allows clients to delete the site's purchase token from DELETE to POST. Unfortunately, it seems that our infrastructure only allows GET and POST methods. I tried to maintain DELETE by sending it in a query parameter called `_method,` but it didn't work. Is there a way to maintain the DELETE method? Is using POST to delete the token valid from our engineering perspective?

File: projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js
Comment (leogermani) [inline projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js:44]: Hi @elliottprogrammer ,

I'm looking only at how you are adding a new parameter to the Redirect URL.

We don't need to add a new parameter in the accepted Args there. The `query` arg is a generic parameter where you can add everything you need.

You have to pass it is as a string. For example: `key1=value1&key2=value2` and this will be added to the final URL after the user reaches the redirect server.

**Review by leogermani — CHANGES_REQUESTED** [review]

File: projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js
Comment (rcanepa) [reply projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js]: Hi @leogermani,

You're right. I replaced the `purchasetoken` query parameter with `query` as you told us and it worked fine.

I pushed a new commit that adds this change and updated the branch against master.

Comment (leogermani) [reply projects/plugins/jetpack/_inc/client/lib/jp-redirect/index.js]: thanks!

**Comment by rcanepa** [comment]
Hey @leogermani, do you think we are ready to merge this PR?

**Comment by rcanepa** [comment]
I forgot to mention that the flow enabled by these changes can be tested fully. If you apply this code to your Jetpack site, you should be able to purchase a Jetpack product as a logged-out user. The start of the flow is the `Plans` tab (see screenshot below).



Here is a demo of the whole flow in action using a test credit card and having `USE_STORE_SANDBOX` enabled on my WPCOM sandbox.

https://user-images.githubusercontent.com/3418513/120356305-300cd480-c2d2-11eb-9bca-27f4b9694dad.mp4

**Comment by leogermani** [comment]
> Hey @leogermani, do you think we are ready to merge this PR?

The code looks good, I'll give it a full round of tests today.

One question: There are other upsell buttons in the Jetpack Settings page, under Ads for example. If you're in site-only connection it will be a "Connect" CTA, but if you have a user connected it's an upsell.

Should this be added to these other links as well?

**Comment by rcanepa** [comment]
> > Hey @leogermani, do you think we are ready to merge this PR?
> 
> The code looks good, I'll give it a full round of tests today.

Awesome!

> One question: There are other upsell buttons in the Jetpack Settings page, under Ads for example. If you're in site-only connection it will be a "Connect" CTA, but if you have a user connected it's an upsell.
> 
> Should this be added to these other links as well?

This is an interesting question because if a site is connected at the user level, one could assume that the user is also logged in to WordPress.com, right? If that is the case, then we don't need to enable this flow to them since it is meant for logged-out users (users that are logged in to WordPress.com will use the default checkout page).

We're in the process of adding a feature that will automatically connect the sites at the user level once a purchase is successfully made. Once that gets done, I think we could replace those "Connect" CTA with upsells that use the logged-out checkout flow because once a user purchases a product, we will automatically connect their site to the new user account. We would get two things from one CTA: a purchase and a connection at the user level. But I would add these changes later in a follow-up PR once the work that enables it gets done.

**Comment by leogermani** [comment]
> This is an interesting question because if a site is connected at the user level, one could assume that the user is also logged in to WordPress.com, right? If that is the case, then we don't need to enable this flow to them since it is meant for logged-out users (users that are logged in to WordPress.com will use the default checkout page).

It's expected that he is registered to WPCOM, but not necessarily logged in at the moment they click on the purchase button.

I asked this because I see that you're adding the purchase token in all cases in this current PR, even when the site is fully connected, so I thought that could be a reason for that. If not, maybe remove the purchase token in those cases?
 
> We're in the process of adding a feature that will automatically connect the sites at the user level once a purchase is successfully made. Once that gets done, I think we could replace those "Connect" CTA with upsells that use the logged-out checkout flow 

Yes! Please let me know if you're tracking this idea somewhere so it's not forgotten, otherwise, I'll add it to our backlog.

> because once a user purchases a product, we will automatically connect their site to the new user account. We would get two things from one CTA: a purchase and a connection at the user level. But I would add these changes later in a follow-up PR once the work that enables it gets done.

I totally agree, there's one gotcha though. There's no plan or checkout experience that highlights "Ads" or "Google Analytics". So today if the user clicks the button to buy these features, they are kind of lost with the options they have as none of them mention the feature they were first interested in (even though they all include it). Not sure if it's a big deal, but definitely something that could be better.

**Comment by rcanepa** [comment]
> It's expected that he is registered to WPCOM, but not necessarily logged in at the moment they click on the purchase button.

Interesting. Well, even in that case, this new logged-out checkout flow doesn't improve the UX for the users that already have a WPCOM account. 

Right now, if those users want to purchase something, they must log in to WordPress.com, and then they can reach the checkout page. The flow for them looks like: `CTA` -> `login page` -> `checkout page`.

Those same users going through the new logged-out checkout flow would see something like: `CTA` -> `checkout page` -> `login page` -> `checkout page`. This happens because the new checkout page allows logged-out users to visit it but they can't log in to their account from it. They must go to the standard login page.

> I asked this because I see that you're adding the purchase token in all cases in this current PR, even when the site is fully connected, so I thought that could be a reason for that. If not, maybe remove the purchase token in those cases?

Yes. Thankfully, you made me realize that this is a mistake. We should not, at least for now, lead users that already have a WPCOM account to this new logged-out checkout flow. They should follow the current one. Is there a frontend utility function we can use to check this?

> > We're in the process of adding a feature that will automatically connect the sites at the user level once a purchase is successfully made. Once that gets done, I think we could replace those "Connect" CTA with upsells that use the logged-out checkout flow
> 
> Yes! Please let me know if you're tracking this idea somewhere so it's not forgotten, otherwise, I'll add it to our backlog.

Bryan started to work on this last week. You can follow the task in 1200152453830945-as-1200329115728364.

> I totally agree, there's one gotcha though. There's no plan or checkout experience that highlights "Ads" or "Google Analytics". So today if the user clicks the button to buy these features, they are kind of lost with the options they have as none of them mention the feature they were first interested in (even though they all include it). Not sure if it's a big deal, but definitely something that could be better.

I agree 100%. I think this is something we will address on the Jetpack Product Pages project.

**Comment by leogermani** [comment]
> Yes. Thankfully, you made me realize that this is a mistake. We should not, at least for now, lead users that already have a WPCOM account to this new logged-out checkout flow. They should follow the current one. Is there a frontend utility function we can use to check this?

In the scope of those links you have `this.props.isLinked` saying whether the current user is connected. (Linked is how we used to refer to a connected user back in the days...)

BUT, this makes me think of something...

This will tell us only if the current user is not connected, but another user could be connected, meaning the current user is not the connection owner. I'm not sure a secondary, not connected, admin should be able to make purchases for the site, while there's another connected admin - the site owner. (well, I just tested it and they can... I didn't go all the way in the flow to understand the implications of that though)

So maybe the right question to ask here is "does this site have a connected owner? (even if not the current user)" and only add the token if not? (not sure)

I'm looking now and I don't think we have this information available on the front end yet (cc @sergeymitr @fgiannar). We might need to add this to the [initial state](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php#L248).

**Comment by rcanepa** [comment]
> This will tell us only if the current user is not connected, but another user could be connected, meaning the current user is not the connection owner. I'm not sure a secondary, not connected, admin should be able to make purchases for the site, while there's another connected admin - the site owner.

I found a utility called `hasConnectedOwner` that might be just what we are looking for https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/client/state/connection/reducer.js#L304.

> (well, I just tested it and they can... I didn't go all the way in the flow to understand the implications of that though)

This shouldn't have happened. The `get_purchase_token` returns a token only when `current_user_can_purchase` returns `true`, and for that to happen, `self::connection()->is_site_connection()` must be `true` as well. My understanding is that `is_site_connection` will only return true if there is no users connected. Is that correct?


Here is a link to `current_user_can_purchase`: https://github.com/Automattic/jetpack/pull/19633/files#diff-028fe0b6a09c6ed2b0a61f21d8db4d21cf8ca35c8904f94f78503619455fe31fR7571-R7584.

**Comment by leogermani** [comment]
> > This will tell us only if the current user is not connected, but another user could be connected, meaning the current user is not the connection owner. I'm not sure a secondary, not connected, admin should be able to make purchases for the site, while there's another connected admin - the site owner.
> 
> I found a utility called `hasConnectedOwner` that might be just what we are looking for https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/_inc/client/state/connection/reducer.js#L304.

There you go!

> 
> > (well, I just tested it and they can... I didn't go all the way in the flow to understand the implications of that though)
> 
> This shouldn't have happened. The `get_purchase_token` returns a token only when `current_user_can_purchase` returns `true`, and for that to happen, `self::connection()->is_site_connection()` must be `true` as well. My understanding is that `is_site_connection` will only return true if there is no users connected. Is that correct?
> 
> Here is a link to `current_user_can_purchase`: https://github.com/Automattic/jetpack/pull/19633/files#diff-028fe0b6a09c6ed2b0a61f21d8db4d21cf8ca35c8904f94f78503619455fe31fR7571-R7584.

Yep, you're right! You got this

File: projects/plugins/jetpack/_inc/client/components/navigation/index.jsx
Comment (rcanepa) [inline projects/plugins/jetpack/_inc/client/components/navigation/index.jsx:56]: Replaced `isLinked` with `hasConnectedOwner` which represents what we really need here.

**Comment by rcanepa** [comment]
Ok! Replaced `isLinked` with `hasConnectedOwner` and updated the PR against master.

**Comment by rcanepa** [comment]
@leogermani, can I have one final review? I would love to merge this PR today if everything is okay with it.

**Review by leogermani — APPROVED** [review]
Looks good to me, thanks for addressing everything.

Just to document for future readers, the Upgrade CTAs in the test instructions are not visible to site-only connected sites, so we're never going to see the purchase tokens there. But we expect to display them in the near future and take advantage of the Logged out checkout flow.

**Pull request merged**
